Improve `find_hung_process.py`

permcody opened the pull request:
- Fixing bug in process trace routine where we were only looking at one trace in the regex
- Adding new --file option to more intuitively load from a cache file
- Automatically finding diverged stacks algorithm based on binary search

closes #7507

permcody committed:
Improve find hung_hung_process.py

- Fixing bug in process trace routine where we were only looking at one trace in the regex
- Adding new `--file` option to more intuitively load from a cache file
- Automatically find diverged stacks algorithm based on binary search

closes #7507

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1441420/) on bc01d2b wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23946/site)

This comment will be updated on new commits.

GiudGiud commented on framework/scripts/find_hung_process.py:
```diff
@@ -231,129 +283,244 @@ def generate_traces(job_num, application_name, num_hosts, num_threads):
+    This class reads a series of stack traces from a file
+    and then processes them to determine if processes in an MPI parallel
+    job have diverged (having differing stack traces that are due to just
+    noise. Memory addresses are stripped.
```
```suggestion
    noise). Memory addresses are stripped.
```

GiudGiud commented on framework/scripts/find_hung_process.py:
```diff
@@ -231,129 +283,244 @@ def generate_traces(job_num, application_name, num_hosts, num_threads):
+        print('\nUnique Stack Traces - frames_kept:', diverged_stack_frame_num)
+        for trace, count in results.items():
+            print(f'{"*"*80}\nCount: {len(count)}\n')
+            if len(count) < 10:
+                print("\n".join(count))
```
just print everything?

permcody replied:
What this logic does is print the hostnames of the outliers. If there are just a few (< 10) hosts in a "unique" stack trace, you'll see those hosts. Otherwise, you'll see just the number of hosts with that trace. It makes a lot of sense (e.g. There are 959 hosts in this stack trace, but host r1i2n3 PID 12345 is in this stace trace).

GiudGiud reviewed (changes requested):
@pbehne please report back once you've used this!

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1454629/) on f25f437 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `framework` coverage rate (required: 81.0%)

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1454629/) on f25f437 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `framework` coverage rate (required: 81.0%)

permcody commented:
@pbehne - I don't suppose you have any reason to test this now? 
@GiudGiud - Any issue with merging? It's a dumb script, it's hard to test but it works. I did talk to Jason about it and we do think we could mock something up later, but none of us have time right now.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1454629/) on f25f437 : invalidated by @permcody

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1454629/) on f25f437 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `framework` coverage rate (required: 81.0%)

GiudGiud commented:
I ll have to try it if @pbehne does not

permcody commented:
It's pretty easy to manufacture a bad test - I have one here: https://gist.github.com/permcody/8c1e8058a994a802b84f426b9321443d

If you want to do it in MOOSE, just add logic anywhere you want that purposely splits the ranks where a parallel operation also occurs. This script should work for any MPI program, not just MOOSE.

pbehne commented:
Sure, I will test it using the above bad test.

pbehne committed:
Added test for find_hung_process.py

The added test case runs find_hung_process.py on a predetermined stack trace
stored in a cache file to find where the stacks diverge.
The results are compared to the expected results.

GiudGiud commented:
nice test.
let s ditch the merge commit

GiudGiud reviewed (dismissed):
lgtm

GiudGiud commented on python/find_hung_process/tests:
```diff
@@ -0,0 +1,9 @@
+    type = RunApp
+    command = '../../framework/scripts/find_hung_process.py -f bad_mpi.4726853.cache'
+    expect_out = "^Count: 7$.*^Count: 1$"
+  []
```
```suggestion
    issues = '#7507'
    design = 'Debug/index.md'
  []
```

GiudGiud commented:
let's squash the last two @pbehne

permcody commented:
He doesn't have maintainer access. I'll have to do it (my branch).
